File: test/Aqua.Tests.VBNet/TypeSystem/ConstructorInfo/When_creating_constructor_info.vb
Imports Aqua.TypeSystem
Imports Aqua.TypeSystem.Extensions
Imports Shouldly
Imports Xunit

Namespace Aqua.Tests.TypeSystem.Extensions
    Public Class When_creating_constructor_info
        Private Class A
        End Class

        <Fact>
        Public Sub Should_create_from_memberinfo()
            Dim a = New A
            Dim typeargs = Array.Empty(Of Type)()
            Dim ctor = a.GetType().GetConstructor(typeargs)
            Dim ctorInfo = New ConstructorInfo(ctor)
            ctorInfo.Name.ShouldBe(".ctor")
            ctorInfo.IsStatic.ShouldBeNull()
            ctorInfo.DeclaringType.ToType().ShouldBeSameAs(a.GetType())
        End Sub

        <Fact>
        Public Sub Should_create_from_name()
            Dim a = New A
            Dim ctorInfo = New ConstructorInfo(".ctor", a.GetType())
            Dim typeargs = Array.Empty(Of Type)()
            Dim ctor = a.GetType().GetConstructor(typeargs)
            ctorInfo.ToConstructorInfo.ShouldBeSameAs(ctor)
        End Sub
    End Class
End Namespace

File: test/Aqua.Tests/TypeSystem/ConstructorInfo/When_creating_constructor_info.cs
// Copyright (c) Christof Senn. All rights reserved. See license.txt in the project root for license information.

namespace Aqua.Tests.TypeSystem.ConstructorInfo;

using Aqua.TypeSystem;
using Shouldly;
using System;
using System.Linq;
using Xunit;
using BindingFlags = System.Reflection.BindingFlags;

public class When_creating_constructor_info
{
    private class A
    {
        static A()
        {
        }

        public A()
        {
        }
    }

    [Fact]
    public void Should_throw_on_creating_by_memberinfo_with_null_parameter()
    {
        Should.Throw<ArgumentNullException>(() => new PropertyInfo((System.Reflection.PropertyInfo)null));
    }

    [Fact]
    public void Should_have_set_is_static_for_type_initializer_created_by_memberinfo()
    {
        var cctor = typeof(A).GetConstructors(BindingFlags.Static | BindingFlags.NonPublic).Single();
        new ConstructorInfo(cctor).IsStatic.ShouldBe(true);
    }

    [Fact]
    public void Should_have_set_is_static_for_type_initializer_created_by_name()
    {
        new ConstructorInfo(".cctor", typeof(A)).IsStatic.ShouldBe(true);
    }

    [Fact]
    public void Should_not_have_set_is_static_by_default_when_created_by_memberinfo()
    {
        var ctor = typeof(A).GetConstructor(Type.EmptyTypes);
        new ConstructorInfo(ctor).IsStatic.ShouldBeNull();
    }

    [Fact]
    public void Should_not_have_set_is_static_by_default_when_created_by_name()
    {
        new ConstructorInfo(".ctor", typeof(A)).IsStatic.ShouldBeNull();
    }
}